Plutarch/Prelude.hs

```diff
@@ -13,17 +14,143 @@ module Plutarch.Prelude (
   (#),
   plet,
   pinl,
-  pcon,
-  pmatch,
   pto,
   pfix,
   Type,
   S,
   PType,
+  PlutusType,
+  PCon (pcon),
+  PMatch (pmatch),
+
+  -- * Integers and integer utilities
+  PInteger,
+  PIntegral (pdiv, pmod, pquot, prem),
+
+  -- * Rational numbers and utilities
+  PRational,
+  pnumerator,
+  pdenominator,
+  pfromInteger,
+  pround,
+
+  -- * Booleans and boolean functions
+  PBool (..),
+  PEq ((#==)),
+  POrd ((#<=), (#<)),
+  pif,
+  pnot,
+  (#&&),
+  (#||),
+
+  -- * Bytestrings and bytestring utilities
+  PByteString,
+  phexByteStr,
+  pconsBS,
+  psliceBS,
+  plengthBS,
+  pindexBS,
+
+  -- * String and string utilities
+  PString,
+  pencodeUtf8,
+  pdecodeUtf8,
+
+  -- * Unit type and utilities
+  PUnit (..),
+
+  -- * Common list typeclass and utilities
+  PListLike (PElemConstraint, pelimList, pcons, pnil, phead, ptail, pnull),
+  PIsListLike,
+  plistEquals,
+  pelem,
+  plength,
+  ptryIndex,
+  pdrop,
+  psingleton,
+  pconcat,
+  pzipWith,
+  pzipWith',
+  pzip,
+  pmap,
+  pfilter,
+  precList,
+  pfoldr,
+  pfoldrLazy,
+  pfoldl,
+  pall,
+  pany,
+
+  -- * Scott encoded list type
+  PList (..),
+
+  -- * Scott encoded maybe type and utilities
+  PMaybe (..),
+
+  -- * Scott encoded either type and utilities
+  PEither (..),
+
+  -- * Scott encoded pair type and utilities
+  PPair (..),
+
+  -- * Builtin types and utilities
+  PData (..),
+  pfstBuiltin,
+  psndBuiltin,
+  PBuiltinPair,
+  PBuiltinList (..),
+  PIsData (pfromData, pdata),
+  PAsData,
+  ppairDataBuiltin,
+  PBuiltinMap,
+
+  -- * DataRepr and related functions
+  PDataFields,
+  PDataRecord,
+  PDataSum,
+  PIsDataRepr,
+  pfield,
+  pletFields,
```

Comment:
We should add the other ones too.

Reply:
Can you give me a quick rundown on the field accessor methods we want to encourage in ideal situations?

Reply:
We want to use the most specific function essentially. If we don't want all the fields, we shouldn't use `pletFields`, and it is very very rare that we want to use all the fields. We probably want a more fine `pletFields` too.

Reply:
Like `pletNFields`? That could be useful more often yeah?

I was under the impression `pletFields` *lazily* binds fields. I suppose that's not the case, though I haven't looked into it much yet.

In any case, we want to encourage `pfield` for sure, correct?

Reply:
I haven't inspected the generated code. It's something we need to ascertain.

Reply:
#177

Reply:
the doc comments in `Example.Field` give a decent rundown

- `pletFields` if (and only if) you want to use all fields
- `pletNFields @n` if you want the 0th field, the nth field and any amount of other fields between
- `pletRangeFields @n @m` if you want the nth field, the mth field and any amount of other fields between
- and `pfield @n` (or a dot accessor if you already have a `PDataRecord` from `pmatch`) for a single field

Reply:
I don't think it's possible for it to be lazy. `plet` is strict after all, and it's not possible for it to know that it can eliminate the variable if it's not used, not to mention that currently it doesn't even *know* when the variable is not used. I think the best solution is to create a variant of `pletFields` that takes a type-level list of fields we want to access.

Reply:
#181

Reply:
I will make a PR after this to change the interface, perhaps without making it efficient. We should have just this in the prelude.

Reply:
I see. For the time being, these are the `DataRepr` related exports I'm thinking of-
```hs
PDataRecord,
PDataSum,
PIsDataRepr,
pfield,
pletRangeFields,
pindexDataRecord,
```
Let me know if any modifications should be made there.

Reply:
>  it is very very rare that we want to use all the fields.

I'd say it's decently common. ScriptContext, for example.

Reply:
Heh, race conditions - I didn't see you guys' comments! I'll wait for Las or Tilde's PR for a streamlined interface and add that to prelude.

Reply:
I don't think you should wait! We'll just change `pletFields` in another PR.

Reply:
The fields will be bound *somewhat* lazily - each list tail is `plet` bound, but the fields (the heads) themselves are not. So the `phead x` terms will not be present in the compiled output unless they are used.